File: 100joon/Gold/2206.py
import sys
from collections import deque

input = sys.stdin.readline

n, m = map(int, input().split())
mat = [list(map(int, input().rstrip())) for _ in range(n)]
visited = [[[0] * 2 for _ in range(m)] for _ in range(n)]

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]


def bfs():
    queue = deque()
    queue.append((0, 0, 0))
    visited[0][0][0] = 1
    while queue:
        x, y, check = queue.popleft()
        if x == n - 1 and y == m - 1:
            print(visited[x][y][check])
            return
        for i in range(4):
            nx = dx[i] + x
            ny = dy[i] + y
            if 0 <= nx < n and 0 <= ny < m:
                if visited[nx][ny][check] == 0:
                    if mat[nx][ny] == 1 and check == 0:
                        visited[nx][ny][1] = visited[x][y][check] + 1
                        queue.append((nx, ny, 1))
                    elif mat[nx][ny] == 0:
                        visited[nx][ny][check] = visited[x][y][check] + 1
                        queue.append((nx, ny, check))

    print(-1)
    return


bfs()

File: 100joon/Gold/2206.js
const input = require("fs").readFileSync("input.txt", "utf-8").split("\n");

class Queue {
  constructor() {
    this.queue = {};
    this.front = 0;
    this.rear = 0;
  }

  push(element) {
    this.queue[this.front++] = element;
  }

  isEmpty() {
    return this.front === this.rear;
  }

  shift() {
    if (this.isEmpty()) {
      console.log("비어있음");
      return;
    }

    const element = this.queue[this.rear++];
    return element;
  }

  size() {
    return this.front - this.rear;
  }
}

const [n, m] = input[0].split(" ");
const mat = Array.from({ length: n }, (_, idx) =>
  input[idx + 1].split("").map(Number)
);
const visited = Array.from({ length: n }, () =>
  Array.from({ length: m }, () => Array(2).fill(0))
);

const dx = [0, 0, -1, 1];
const dy = [-1, 1, 0, 0];
let ans = -1;
const queue = new Queue();
queue.push([0, 0, 0, 1]);
while (queue.size()) {
  const [x, y, z, cnt] = queue.shift();
  if (x === n - 1 && y === m - 1) {
    ans = cnt;
    break;
  }

  for (let i = 0; i < 4; i++) {
    const nx = x + dx[i];
    const ny = y + dy[i];
    if (0 <= nx && nx < n && 0 <= ny && ny < m) {
      // 이동 가능
      if (mat[nx][ny] === 0 && !visited[nx][ny][z]) {
        queue.push([nx, ny, z, cnt + 1]);
        visited[nx][ny][z] = 1;
      } else if (mat[nx][ny] === 1 && z === 0 && !visited[nx][ny][z + 1]) {
        queue.push([nx, ny, z + 1, cnt + 1]);
        visited[nx][ny][z + 1] = 1;
      }
    }
  }
}

console.log(ans);
